exec/java-exec/src/main/java/org/apache/drill/exec/planner/logical/partition/PruneScanRule.java

```diff
@@ -143,6 +144,11 @@ public static final RelOptRule getDirFilterOnScan(OptimizerRulesContext optimize
   }
 
   protected void doOnMatch(RelOptRuleCall call, Filter filterRel, Project projectRel, TableScan scanRel) {
+    if (wasAllPartitionsPruned) {
```

Comment:
Agree that this should not be an internal state.  I had added it to overcome an issue where the prunescan rule gets re-applied even when there are not qualifying partitions (this happens because we always populate the newPartitions list with at least 1 entry to produce a valid schema during execution).   

I tried removing this flag but will need some associated changes to prevent the rule from being reapplied without termination.  The old code (without my patch) terminates the pruning because FileSelection.supportDirPruning() returns FALSE when the selection has directories.   Whereas, for this optimization I want to do the pruning on directories first.   So, let me think of a way to address that.